### ISSUE TITLE
Symbol not found: _caml_builtin_cprim

### ISSUE THREAD

**Issue by dhil** [ocamllabs/ocaml-multicore#124]
Let `test.ml` be
```
print_endline (Str.quote "asdf")
```
then on MacOS X (El Capitan) with a locally checked out up-to-date trunk version of Multicore OCaml `ocaml.opt` compiler I get 
```
$ ocamlc.opt -verbose -o test.byte str.cma test.ml
File "test.ml", line 1:
Error: Error on dynamically loaded library: /Users/dhil/.opam/4.02.2+local-git-master/lib/ocaml/stublibs/dllcamlstr.so: dlopen(/Users/dhil/.opam/4.02.2+local-git-master/lib/ocaml/stublibs/dllcamlstr.so, 138): Symbol not found: _caml_builtin_cprim
  Referenced from: /Users/dhil/.opam/4.02.2+local-git-master/lib/ocaml/stublibs/dllcamlstr.so
  Expected in: flat namespace
 in /Users/dhil/.opam/4.02.2+local-git-master/lib/ocaml/stublibs/dllcamlstr.so
```
While compiling with `ocamlc` seems to work just fine
```
$ ocamlc -verbose -o test.byte str.cma test.ml
$ echo $?
0
```

**Comment by stedolan** [comment]
This works fine for me on Linux. @kayceesrk could you try to reproduce this on OS X?

**Comment by dhil** [comment]
Yes, I can confirm it works fine on my Linux (Ubuntu 16.04) machine as well.

**Comment by kayceesrk** [comment]
I can reproduce this error in OSX.

**Comment by dhil** [comment]
For reference it works fine on Mac OS X with the vanilla OCaml compiler (tested with 4.04.0).

**Comment by dhil** [comment]
For reference: The problem seems to be the following rule in `/otherlibs/Makefile.shared`:
```Makefile
lib$(CLIBNAME).$(A): $(COBJS)
	$(MKLIB) -oc $(CLIBNAME) $(COBJS) $(BYTECAMLRUN) $(LDOPTS)
```
it shouldn't include `$(BYTECAMLRUN)`.

**Comment by kayceesrk** [comment]
If you simply remove `$(BYTECAMLRUN)` compilation on osx is likely to break completely. This was added to avoid shared lib support on osx. But @stedolan tells me this is an issue that has been fixed in the wrong place. Essentially, in the current code, pthread local variables (`__thread`) annotations may be missing on certain variables in the header files. If each one of those is properly annotated as thread-local, then the linker would stop complaining. That would fix the problem.

**Comment by kayceesrk** [comment]
Fixed by https://github.com/ocamllabs/ocaml-multicore/commit/167e64b94ae2516e4620b6ad2868eb3fce55e77a. @dhil can you confirm?

**Comment by dhil** [comment]
> Fixed by 167e64b. @dhil can you confirm?

Yes, it works on my end too.
